### PR TITLE
[SPARK-22032][PySpark] Speed up StructType conversion

### PR DESCRIPTION
**Pull request by maver1ck**
## What changes were proposed in this pull request?

StructType.fromInternal is calling f.fromInternal(v) for every field.
We can use precalculated information about type to limit the number of function calls. (its calculated once per StructType and used in per record calculations)

Benchmarks (Python profiler)
```
df = spark.range(10000000).selectExpr("id as id0", "id as id1", "id as id2", "id as id3", "id as id4", "id as id5", "id as id6", "id as id7", "id as id8", "id as id9", "struct(id) as s").cache()
df.count()
df.rdd.map(lambda x: x).count()
```

Before
```
310274584 function calls (300272456 primitive calls) in 1320.684 seconds

Ordered by: internal time, cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 10000000  253.417    0.000  486.991    0.000 types.py:619(<listcomp>)
 30000000  192.272    0.000 1009.986    0.000 types.py:612(fromInternal)
100000000  176.140    0.000  176.140    0.000 types.py:88(fromInternal)
 20000000  156.832    0.000  328.093    0.000 types.py:1471(_create_row)
    14000  107.206    0.008 1237.917    0.088 {built-in method loads}
 20000000   80.176    0.000 1090.162    0.000 types.py:1468(<lambda>)
```

After
```
210274584 function calls (200272456 primitive calls) in 1035.974 seconds

Ordered by: internal time, cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
 30000000  215.845    0.000  698.748    0.000 types.py:612(fromInternal)
 20000000  165.042    0.000  351.572    0.000 types.py:1471(_create_row)
    14000  116.834    0.008  946.791    0.068 {built-in method loads}
 20000000   87.326    0.000  786.073    0.000 types.py:1468(<lambda>)
 20000000   85.477    0.000  134.607    0.000 types.py:1519(__new__)
 10000000   65.777    0.000  126.712    0.000 types.py:619(<listcomp>)
```

Main difference is types.py:619(<listcomp>) and types.py:88(fromInternal) (which is removed in After)
The number of function calls is 100 million less. And performance is 20% better.

Benchmark (worst case scenario.)

Test
```
df = spark.range(1000000).selectExpr("current_timestamp as id0", "current_timestamp as id1", "current_timestamp as id2", "current_timestamp as id3", "current_timestamp as id4", "current_timestamp as id5", "current_timestamp as id6", "current_timestamp as id7", "current_timestamp as id8", "current_timestamp as id9").cache()
df.count()
df.rdd.map(lambda x: x).count()
```

Before
```
31166064 function calls (31163984 primitive calls) in 150.882 seconds
```

After
```
31166064 function calls (31163984 primitive calls) in 153.220 seconds
```

IMPORTANT:
The benchmark was done on top of https://github.com/apache/spark/pull/19246.
Without https://github.com/apache/spark/pull/19246 the performance improvement will be even greater.

## How was this patch tested?

Existing tests.
Performance benchmark.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
How many functions call does it save and much it improves? I'd not bother fixing this.

**Comment by HyukjinKwon** [comment]
To be honest, it looks too trivial that I won't bother.

**Comment by maver1ck** [comment]
I was checking this with my production code.
This give me about 6-7% of speed up and remove 408 millions of function calls :)

I'll try to create benchmark for this.

**Comment by HyukjinKwon** [comment]
Did it save 6~7% of the total execution time?

**Comment by maver1ck** [comment]
Yep.

**Comment by HyukjinKwon** [comment]
Okay, then let's go ahead then. Let'd add some numbers in the PR description.

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -619,7 +621,8 @@ def fromInternal(self, obj):
             # it's already converted by pickler
             return obj
         if self._needSerializeAnyField:
-            values = [f.fromInternal(v) for f, v in zip(self.fields, obj)]
+            values = [f.fromInternal(v) if n else v
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: Ah, I see. This can be recursive and per-record and we avoid here by pre-computing. I see. That makes much sense.

Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Let's describe this in more details and add some numbers (and in your other PRs too).

**Comment by maver1ck** [comment]
I added benchmark for this code

**Comment by gatorsmile** [comment]
Could you mark [PySpark] in the title? cc @ueshin

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: Could we run a benchmark with the worst case, when all columns are needed to be converted?  I think here we pay another if and extra element in the zip to prevert function call basically. This one looks okay practically but I guess we should also identify the downside.

Also, let's add a comment here to describe what we are doing here and also add some links to this PR for other guys to refer the benchmarks.

Comment (maver1ck) [reply python/pyspark/sql/types.py]: I'll add one more optimization here.
And then I'll do benchmarks.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -483,7 +483,8 @@ def __init__(self, fields=None):
             self.names = [f.name for f in fields]
             assert all(isinstance(f, StructField) for f in fields),\
                 "fields should be a list of StructField"
-        self._needSerializeAnyField = any(f.needConversion() for f in self)
+        self._needConversion = [f.needConversion() for f in self]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py]: I'd rename this to another, for example, `_needConversions` (or others if there are) and leave a comment here why we do this.

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Thanks for testing this out. Let's add a comment here to explain what we are doing and move the worst case benchmark into the PR description.

**Review by HyukjinKwon — APPROVED** [review]
Minimal change and practically significant improvement. LGTM. @ueshin, do you maybe have some comments on this?

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (viirya) [inline python/pyspark/sql/types.py]: Can we use the similar trick on `toInternal`?

Comment (HyukjinKwon) [reply python/pyspark/sql/types.py]: Yea, looks we could too.

Comment (maver1ck) [reply python/pyspark/sql/types.py]: Done.

**Comment by viirya** [comment]
LGTM

**Comment by HyukjinKwon** [comment]
LGTM too but hey @maver1ck could you add some comments around the codes and move the worst case benchmarks into the PR description? I guess this wouldn't be too demanding.

**Comment by maver1ck** [comment]
Done.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by ueshin** [comment]
A late LGTM. Btw, can we use the same idea for `MapType`?

**Comment by HyukjinKwon** [comment]
Thanks for double checking @ueshin.

Yes, I noticed that too while reviewing it. I just decided to merge it as is because I am quite sure of this one given struct type is the root type and this case looks quite common, and regarding that it looks the first contribution. Even though this one has a downside, practically the improvement looked better.

I am also fine with doing this for others too (I am +0 for other types).

**Comment by maver1ck** [comment]
@ueshin 
I think that for Maptype this is not a solution because every key / value of MapType is the same type so we need conversion for all entries or for nothing

**Comment by HyukjinKwon** [comment]
We can split two `needConversion` for key and value only and save key conversion or value conversion call though?

**Pull request closed without merging**